### ISSUE TITLE
Beat and bassline bug

### ISSUE THREAD

**Issue by AbaddonBent** [LMMS/lmms#5673]
There's a bug where a instrument in beat and bassline has automation and the tracks above that instrument become hidden and un-editable.
![image](https://user-images.githubusercontent.com/69567825/92424486-5e8dc280-f139-11ea-82a2-52292339ffe0.png)

I was making pitch adjustments when I noticed this

**Comment by thmueller64** [comment]
Looking at your screenshot I assume you are using the 1.2.2 Appimage.
I unsuccessfully tried to reproduce the bug using the information in your screenshot with the following steps:
  1. Create a beat/bassline instance
  2. Create patterns inside the track with audio samples only (I created 16 as above)
  3. Automate the pitch of one of the last patterns

Can you reliably reproduce the bug? If so, could you please detail the steps to reproduce it? You could also share a project file.

**Comment by zonkmachine** [comment]
I think this is mentioned somewhere else. A method to make empty Beat/Bassline tracks is to delete all beat bassline instances in the Song Editor. Existing tracks in the Beat/Bassline will then go blank. Sounds like you managed to do it in some other way.

**Comment by AbaddonBent** [comment]
I've been trying to duplicate it but nothing has been showing....
The file I had has changed and when I opened it, it has reset the beat and bassline, the screenshot is the only proof I've had, but I'll keep trying to see if it happens.
My theory is that it was a one time thing

**Comment by AbaddonBent** [comment]
What I remembered what I did when this happened, I accidentally pressed the add sample track button. Then removed it and pressed the automation track and figured I'd mess with the pitch

**Comment by IanCaio** [comment]
I have experienced something similar by doing the following steps (I described it on my PR because at first I thought it could be related to it but then realized it was not):

1. Open the default project (TripleOsc, SampleTrack, BBTrack and AutomationTrack).
2. Create 2 TCOs on the TripleOsc track on Bars 1 and 2
3. Ctrl+Drag the TripleOsc track to the BBEditor
4. Add some notes to the Kicker track (not necessary for the bug to happen)
5. Clone the BBTrack

Could it be you did something similar, like dragging an instrument that already had TCOs to the BB Editor before cloning? I'm still not sure what causes this bug, but could anyone check if they can reproduce it by following those steps?

**Comment by IanCaio** [comment]
Afaik, all BB Tracks share the same tracks, and what differs a BB pattern from another is the TCO being played/visualized. Something like this:

```
          BBTrack1        BBTrack2       BBTrack3
Kicker: === TCO 1 === | === TCO 2 === | === TCO 3 ===
Snare:  === TCO 1 === | === TCO 2 === | === TCO 2 ===
```

Trying to do some debugging I found out that the Kicker track on the default project had 2 TCOs instead of 1, which I'd expect since the default project only has 1 BBTrack at first. I think that could be related to the bug, but that's as far as I went into investigating it.

**Comment by IanCaio** [comment]
Could everyone that managed to experience this bug try to reproduce it on the build from PR #5699 ? I think I accidentally found the reason behind the bug and fixed it there. The method that creates TCOs expects a position as an argument but in many classes it didn't actually place the TCO on the position given. That went unnoticed everywhere else because every time a TCO was created there was an explicit call to `movePosition` right after it, but on the method `getTCO` there wasn't one, so the TCO was expected to be created in a position but wasn't. That's probably why they are all missing in the BBEditor, they are misplaced.

That PR should fix it but it needs testing.

**Comment by AbaddonBent** [comment]
I got the issue again, this time I recorded it!
I don't know what happened but as soon as I added another track that uses the same instruments and up, the original sampled instrument and up is gone, meaning that the beat and bassline can't see it! on those tracks, I right clicked and edited them so my thought is that when you have an instrument and it uses custom edits such as odd timing, it'll bug out and appear like nothing was there.
here's the video
[got em.zip](https://github.com/LMMS/lmms/files/5498439/got.em.zip)

here's the project file
[grim theme.zip](https://github.com/LMMS/lmms/files/5498434/grim.theme.zip)

I had to zip them because github don't allow mp4/blend files

**Comment by AbaddonBent** [comment]
I reloaded the file and it automatically solved it?

**Comment by Spekular** [comment]
@IanCaio should this be closed now that #5699 is merged?

**Comment by IanCaio** [comment]
@Spekular I believe so, the only steps I had to reliably reproduce this bug now work properly with the `createTCO` method fixes, so I think misplacement of the TCOs on the BBTrack was the root cause. Closing it then! :smiley:
